www/class/config-generate/host.class.php

```diff
@@ -132,6 +135,249 @@ private function getServicesByHg(&$host)
+            }
+        }
+        return array();
+    }
+
+    /**
+     * @param array $host
+     * @param string $attribute
+     * @param $attributeAdditive
+     * @return array|int|mixed
+     */
+    private function manageVerticalInheritance(array &$host, string $attribute, $attributeAdditive)
+    {
+        $results = $host[$attribute . '_cache'];
+        if (count($results) > 0
+            && (is_null($host[$attributeAdditive]) || $host[$attributeAdditive] != 1)) {
+            return $results;
+        }
+        
+        $hostsTpl = HostTemplate::getInstance($this->dependencyInjector)->hosts;
+        $hostIdCache = null;
+        foreach ($host['htpl'] as $hostIdTopLevel) {
+            $computedCache = array();
+            if (!isset($hostsTpl[$hostIdTopLevel][$attribute . '_computed_cache'])) {
+                $stack = array(array($hostIdTopLevel, 1));
+                $loop = array();
+                $currentLevelCatch = null;
+                while ((list($hostId, $level) = array_shift($stack))) {
+                    if (!is_null($currentLevelCatch) && $currentLevelCatch >= $level) {
+                        break;
+                    }
+                    if (isset($loop[$hostId])) {
+                        continue;
+                    }
+                    $loop[$hostId] = 1;
+
+                    if (!is_null($hostsTpl[$hostId]['notifications_enabled'])
+                        && (int)$hostsTpl[$hostId]['notifications_enabled'] === 0) {
+                        continue;
+                    }
+
+                    if (count($hostsTpl[$hostId][$attribute . '_cache']) > 0) {
+                        $computedCache = array_merge($computedCache, $hostsTpl[$hostId][$attribute . '_cache']);
+                        $currentLevelCatch = $level;
+                        if (is_null($hostsTpl[$hostId][$attributeAdditive]) || $hostsTpl[$hostId][$attributeAdditive] != 1) {
+                            break;
+                        }
+                    }
+
+                    foreach (array_reverse($hostsTpl[$hostId]['htpl']) as $htplId) {
+                        array_unshift($stack, array($htplId, $level + 1));
+                    }
+                }
+
+                $hostsTpl[$hostIdTopLevel][$attribute . '_computed_cache'] = array_unique($computedCache);
+            }
+
+            if (count($hostsTpl[$hostIdTopLevel][$attribute . '_computed_cache']) > 0) {
+                $hostIdCache = $hostIdTopLevel;
+                break;
+            }
+        }
+
+        if (!is_null($hostIdCache)) {
+            $results = array_unique(
+                array_merge($results, $hostsTpl[$hostIdCache][$attribute . '_computed_cache']),
+                SORT_NUMERIC
+            );
+        }
+        return $results;
+    }
+
+    /**
+     * @param array $host
+     * @param array|null $cg
+     */
+    private function setContactGroups(array &$host, ?array $cg) : void
```

Comment:
This array is used in a foreach loop so it cannot be null
```suggestion
    private function setContactGroups(array &$host, array $cg = []) : void
```